src/index.js

```diff
@@ -14,34 +14,64 @@ const _DNS = or(
 )
 
 const IP = or(base('ip4'), base('ip6'))
-const TCP = and(IP, base('tcp'))
-const UDP = and(IP, base('udp'))
-const UTP = and(UDP, base('utp'))
 
-const DNS = or(
+const _UDP = and(IP, base('udp'))
+const UDP = or(
+  and(_UDP, base('ipfs')),
```

Comment:
Without wanting to go back and forth. Is there any reason why a transport code itself wouldn't be able to make two assertions .matchIPFS and .matchUDP so that if the first one happen, it decapsulates IPFS first and then does the matchUDP?

Note, that is how all the transports have been working:
- https://github.com/libp2p/js-libp2p-websockets/blob/master/src/index.js#L46-L57
- https://github.com/libp2p/js-libp2p-tcp/blob/master/src/index.js#L66-L76

There is a separation between what is a valid addr for a spec transport (i.e TCP was here before) and there is the addrs that our libp2p-<insert transport name> accept.

libp2p-tcp accepts TCP multiaddrs or TCP + IPFS, but that assertion should happen on the filter function and not in mafmt, which is a utility to check capabilities/presence of certain addrs.

Reply:
Basically because of this - https://github.com/libp2p/js-libp2p-tcp/pull/80/files, we though at the time that having mafmt handle those special circuit cases such as `/ip4/127.0.0.1/tcp/4001/ws/ipfs/QmRelay/p2p-circuit/ipfs/QmDst` would be a better approach. I'm not so sure anymore, I think the initial approach of handling the special cases might be simpler and safer. 

The two relevant PR that lead to this change are - 
- https://github.com/libp2p/js-libp2p-websockets/pull/61
- https://github.com/libp2p/js-libp2p-tcp/pull/80

Reply:
Also, I think that since then, we've also scratched listening on circuit addresses and the more advanced circuit-relay features - we might not even need this anymore (for now) because we're not going to be using that kind of addressing for now. I would say we leave it alone for now and revisit when we actually need this feature. 

Glad you asked the question tho!

Reply:
@diasdavid seems like I spoke too soon, we still need this as we are supporting specific relay addrs. We can go with this approach or reopen the PRs linked above (handle circuit as a special case in `.filter`). I'm not so sure of which one is better. This PR is definitely higher risk as it affects everything that uses mafmt, handling a special case in `.filter` is uglier, but might be safer for the time being?

Here is @vyzo response regarding Go's handling of specific circuit addrs.

```
dryajov: it accepts them
10:13 but unless it's an active relay it's pretty much a hint
10:13 if it's an active relay, then it adds the address to its peerstore for the dial attempt
10:13 so it will honor them
10:14 ah,the dialer
10:14 sorry, i was mislieading
10:14 it handles them just fine
10:14 it adds to the peerstore for the dial attempt
10:14 so it's a hint
10:14 may or may not be used by the dialer
10:14 and maybe there is already an existing connection
```

Reply:
After a conversation with @diasdavid we've decided to handle this in the `filter` function of each transport.